NEXT-23682 - Add varnish xkey support (#724)
shopware/docs: guides/hosting/infrastructure/reverse-http-cache.md

}
```

### Using Varnish XKey Module without Redis

Varnish XKey is a cache key module that allows you to use Varnish with surrogate keys. It is a module that is not included in the default Varnish installation. It is available for Varnish 4.1 and 6.0.

The module is available for download on [GitHub](https://github.com/varnish/varnish-modules/blob/master/src/vmod_xkey.vcc)

{% hint style="warning" %}
This feature has been introduced with Shopware version 6.4.17.0
{% endhint %}

And also needs to be enabled in the `config/packages/shopware.yml` file:

```yaml
storefront:
  reverse_proxy:
    enabled: true
    use_varnish_xkey: true
    hosts:
      - 'varnish-host'
```

Varnish Config:

```vcl
vcl 4.0;

import std;
import xkey;

> # You should specify here all your app nodes and use round robin to select a backend

Reviewer: ```suggestion
# Specify your app nodes here. Use round-robin balancing to add more than one.
```

backend default {
    .host = "<app-host>";
    .port = "80";
}

# ACL for purgers IP. (This needs to contain app server ips)
acl purgers {
    "127.0.0.1";
    "localhost";
    "::1";
}

sub vcl_recv {
    # Mitigate httpoxy application vulnerability, see: https://httpoxy.org/
    unset req.http.Proxy;

>     # Strip query strings only needed by browser javascript. Customize to used tags.

Reviewer: ```suggestion
    #  Ignore query strings that are only necessary for the js on the client.  Customize as needed.
```

    if (req.url ~ "(\?|&)(pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid)=") {
        # see rfc3986#section-2.3 "Unreserved Characters" for regex
        set req.url = regsuball(req.url, "(pk_campaign|piwik_campaign|pk_kwd|piwik_kwd|pk_keyword|pixelId|kwid|kw|adid|chl|dv|nk|pa|camid|adgid|cx|ie|cof|siteurl|utm_[a-z]+|_ga|gclid)=[A-Za-z0-9\-\_\.\~]+&?", "");
    }
    set req.url = regsub(req.url, "(\?|\?&|&)$", "");

    # Normalize query arguments
    set req.url = std.querysort(req.url);

    # Make sure that the client ip is forward to the client.
    if (req.http.x-forwarded-for) {
        set req.http.X-Forwarded-For = req.http.X-Forwarded-For + ", " + client.ip;
    } else {
        set req.http.X-Forwarded-For = client.ip;
    }

    # Handle PURGE
    if (req.method == "PURGE") {
        if (client.ip !~ purgers) {
            return (synth(403, "Forbidden"));
        }
        if (req.http.xkey) {
            set req.http.n-gone = xkey.purge(req.http.xkey);

>             # Replace this line for

Reviewer: Replace for what?

>             # Replace this line for

Author: ```suggestion
            # To enable soft-purge replace this line with above
```
